### ISSUE TITLE
dolt gives constraint violation on merge when one doesn't exist

### ISSUE THREAD

**Issue by ricardoreiter** [dolthub/dolt#6319]
This was a tricky one to isolate the repro steps. I know that updating the row key isn't recommended as dolt treats it as a delete/add, it's a weird scenario but it happens in our DB for now.

Repro:
```
dolt sql -q "create table tableA (keyC varchar(255) not null, col1 varchar(255), col2 varchar(255), col3 varchar(255), dataA varchar(255), PRIMARY KEY (keyC), UNIQUE KEY col1_col2_col3_unique (col1, col2, col3))"
dolt sql -q "insert into tableA (keyC, col1, col2, col3, dataA) values ('key-a', '1', '2', '3', 'test1')"
dolt add .
dolt commit -m "new table"

dolt checkout -b feature

dolt checkout main
dolt sql -q "update tableA set keyC = 'key-b'"
dolt add .
dolt commit -m "update row"

dolt checkout feature
dolt sql -q "insert into tableA (keyC, col1, col2, col3, dataA) values ('key-c', '2', '2', '3', 'test3')"
dolt add .
dolt commit -m "new row in branch"
```

then 
```
dolt merge main

Updating ceqmfi25vseh471acp1ek7juff02atpg..28jmdrac8eq9le7vk5eukda9bltnt9pn
Auto-merging tableA
CONSTRAINT VIOLATION (content): Merge created constraint violation in tableA
Automatic merge failed; 1 table(s) are unmerged.
Fix constraint violations and then commit the result.
Constraint violations for the working set may be viewed using the 'dolt_constraint_violations' system table.
They may be queried and removed per-table using the 'dolt_constraint_violations_TABLENAME' system table.
```

```
dolt sql -q "select * from dolt_constraint_violations_tableA"
+----------------------------------+----------------+-------+------+------+------+-------+------------------------------------------------------------------------+
| from_root_ish                    | violation_type | keyC  | col1 | col2 | col3 | dataA | violation_info                                                         |
+----------------------------------+----------------+-------+------+------+------+-------+------------------------------------------------------------------------+
| 28jmdrac8eq9le7vk5eukda9bltnt9pn | unique index   | key-a | 1    | 2    | 3    | test1 | {"Name": "col1_col2_col3_unique", "Columns": ["col1', 'col2', 'col3"]} |
| 28jmdrac8eq9le7vk5eukda9bltnt9pn | unique index   | key-b | 1    | 2    | 3    | test1 | {"Name": "col1_col2_col3_unique", "Columns": ["col1', 'col2', 'col3"]} |
+----------------------------------+----------------+-------+------+------+------+-------+------------------------------------------------------------------------+

dolt sql -q "select * from tableA"
+-------+------+------+------+-------+
| keyC  | col1 | col2 | col3 | dataA |
+-------+------+------+------+-------+
| key-b | 1    | 2    | 3    | test1 |
| key-c | 2    | 2    | 3    | test3 |
+-------+------+------+------+-------+
```

**Comment by timsehn** [comment]
I was able to repro using `dolt merge main` but not checking out main and merging feature. Something about the primary key change definitely messes us up. I hope we're updating the unique index in that case.

**Comment by fulghum** [comment]
I took a quick look at this in the debugger last night, and we do see the diff events come in as a deletion of row (id=1) followed by a diff event with an insert of row (id=2). The issue with the unique validator seems to be that it grabs the row data before any of the diff events come in, then it always validates uniqueness against that version of the data, without ever processing the deletes (those get processed in a different part of the code that does not update the view the unique validator has). 

I have a couple of ideas to try out to get this working, although there may be some trickiness with the ordering of how the events come in. In this case, since the delete comes in first, we should be able to support this fairly easily, but I also want to explore some other cases with different ordering to see if the same issue can be triggered.

**Comment by fulghum** [comment]
As expected, there is some trickiness with the diff event ordering. I have a prototype that fixes the repro script above, however it works because the row deletion event for pk=`key-a` shows up *before* the row insert event for pk=`key-b`. Swapping these values causes the insert to show up first, which will trigger the incorrect constraint violation behavior above. 

I've got a few ideas on how to process the deletes first and will try out some of those next.

**Comment by fulghum** [comment]
Another quick update... I've been exploring a few ideas to make this work in all cases. The prototype I mentioned above only works in some cases, and I want to solve this problem generally for you of course, and not only when the edits happen to order the data in a certain way. 

One idea is to process all the row deletion events first, and then re-iterate through the diffs a second time after that. I think that _could_ work, but it's a bit clumsy to process the diffs twice. Max had a good idea to instead update another data structure directly to try and remove violation information when we see row deletes and I've been testing that idea out. It looks like it's going to work, and is much more efficient than re-iterating over diffs. I still need to finish a little bit more logic and do some more testing to validate that this is really going to work, but it's looking promising.

**Comment by fulghum** [comment]
More progress on this today. The approach I'm taking is a little trickier than I expected (e.g. have to jump through three different key formats: primary -> secondary -> artifactMap), but I think it's still worth it to avoid having to replay the diff events. As of this afternoon, I have both tests cases working: one where the row deletion event shows up first, and one where the row insert event shows up first. 

Next, I need to tidy up the code and test with some more edge cases to see if there are any other gaps I can find.

**Comment by fulghum** [comment]
Thanks for helping us find this issue @ricardoreiter! 🙏  I've got a fix merged in and will run a release later today for you.

**Comment by fulghum** [comment]
This fix is released in [Dolt 1.7.6](https://github.com/dolthub/dolt/releases/tag/v1.7.6) and is available to upgrade to on Hosted Dolt. Let us know if you hit any issues with it!
